.github/workflows/check-api.yml

```diff
@@ -0,0 +1,93 @@
+
+on:
+  push:
+    paths-ignore:
+      - "doc/**"
+      - "**/README.md"
+      - "CONTRIBUTING.md"
+    branches:
+      - master
+  pull_request:
+    paths-ignore:
+      - "doc/**"
+      - "**/README.md"
+      - "CONTRIBUTING.md"
+  release:
+    types:
+      - created
+  workflow_call:
+
+jobs:
+  check-api:
+    name: Check API using ${{ matrix.sys.os }} ghc-${{ matrix.ghc }}
+    runs-on: ${{ matrix.sys.os }}
+    strategy:
+      matrix:
+        # we check API only on one platform and ghc release, since it shouldn't
+        # vary elsewhere and the API tracer is sensitive to both
+        sys:
+          - { os: ubuntu-latest }
+        ghc:
+          [
+            # print-api only supports a small subset of ghc versions
+            "9.10.1",
+          ]
+
+    steps:
+
+      - uses: actions/checkout@v4
+
+      - uses: haskell-actions/setup@v2
+        id: setup-haskell
+        with:
+          ghc-version: ${{ matrix.ghc }}
+          cabal-version: 3.12.1.0 # see https://github.com/haskell/cabal/pull/10251
+          ghcup-release-channel: https://raw.githubusercontent.com/haskell/ghcup-metadata/master/ghcup-prereleases-0.0.8.yaml
+
+      # I was going to use the canned action, but it only supports a single package and reinstalls the same binary each time
+      - name: Install print-api
+        run: |
+          wget -q https://github.com/Kleidukos/print-api/releases/download/v0.1.0.1/print-api-0.1.0.1-Linux-static-${{ matrix.ghc }}-x86_64.tar.gz
+          tar -xzvf print-api-0.1.0.1-Linux-static-${{ matrix.ghc }}-x86_64.tar.gz
+          chmod +x print-api
+
+      # print-api needs environment files. It also doesn't make a lot of sense to use the cached builds, sadly,
+      # since they're special in different ways (bootstrap and validate) and we want a vanilla build. And there
+      # isn't enough cache space to make a third cache, even though this is a very limited build.
+      - name: Build Cabal with environment files
+        run: cabal build Cabal-syntax Cabal --write-ghc-environment-files=always
+
+      - name: Generate Cabal-syntax and Cabal APIs
+        run: |
+          ./print-api --package-name Cabal-syntax > Cabal-syntax-${{ matrix.ghc }}-${{ matrix.sys.os }}.api
+          ./print-api --package-name Cabal > Cabal-${{ matrix.ghc }}-${{ matrix.sys.os }}.api
+
+      # for convenience, since large changes would be a pain to reconstruct from diffs and
+      # contributors aren't guaranteed to have ubuntu-latest handy
+      # run this _before_ checking the API, so hopefully they're available on failure
+      # because that's the point of making them artifacts!
+      - uses: actions/upload-artifact@v4
+        with:
+          name: Cabal-api
+          path: '*.api'
+
+      - name: Check Cabal-syntax and Cabal APIs
+        run: |
+          rc=0
+          if cmp -s Cabal-syntax/Cabal-syntax-${{ matrix.ghc }}-${{ matrix.sys.os }}.api Cabal-syntax-${{ matrix.ghc }}-${{ matrix.sys.os }}.api; then
```

Comment:
I feel like the decision to opt for a simpler check with `cmp` rather than going full`diff` is not the right choice for CI in particular. You say: 

> switched out diff for cmp, because with Cabal's 1MB+ API footprint nobody really cares about the details, just that it changed. 

So is the only reason for not doing diff is that no one cares?  I think I may care in some reasons. Is there a more objective reason? Does it save time or something? One could run `cmp` first and only if that fails run `diff`...

Reply:
Maybe I'm overreacting to the first run, where it was `diff`ing against an empty file and so dumped the entire API. It was pretty much impenetrable, unorganized, and managed to break GitHub's log viewer so I ended up downloading the build log to read it.

Some of this is just that `print-api` output isn't really designed for humans. Much of it is that we just have too much API, in particular in `Cabal`. (Diffs work for `text-display`, but `print-api`'s entire output is about 12 lines there. `Cabal`'s is 1000 times larger, even if you remove its re-export of `Cabal-syntax`.)

In re "run `cmp` first and only if that fails run `diff`", both produce no output if they succeed so there would be no difference. Possibly we could run `diff` to a file and output it if it's small enough, but then we will bikeshed over what counts as "small enough".

Reply:
I've switched to `diff` (which shortly will be `diff -c` because plain `diff` output is a little too obtuse) and show the changes if there's fewer than 50 lines. We can tune this to some extent, likely limited by what GitHub's log viewer can cope with.